### PR TITLE
Add new web framework and re-implement existing REST APIs

### PR DESCRIPTION
**Pull request by stringlytyped**
This PR introduces a light web framework on top of Tornado in order to impose a standard organisational pattern to Keylime's REST API code. This is in service of providing a solid foundation on which to implement the push model (https://github.com/keylime/enhancements/issues/103). Additionally, this PR aims to re-implement the existing pull APIs in the new framework to ensure backwards compatibility.

Only the registrar's REST APIs are included in this PR; verifier changes will follow once I get some feedback. This PR is already pretty substantial, so I don't want to make it even more overwhelming.

The changes so far are organised according to the following directory structure:

- `keylime/web`: contains all web-layer code
- `keylime/web/base`: contains the class library used by other files in the `web` directory
- `keylime/web/registrar`: contains the _controllers_ for the registrar (currently consists of a single `agents_controller.py`)
- `keylime/web/registrar_server.py`: the main entry point for incoming requests to the registrar which are then routed to the appropriate controller
- `keylime/models`: contains all data model-layer code
- `keylime/models/base`: contains the class library used by other files in the `models` directory
- `keylime/models/base/type.py` and `keylime/models/base/types`: handles incoming/outgoing model data according to its declared type
- `keylime/models/registrar`: contains the models relevant to the registrar (currently consists of a single `registrar_agent.py` model)

If anyone has any questions about the main building blocks which I have defined (servers, routes, controllers and models), or anything else really, let me know and I'll do my best to clarify.

### CODE REVIEW

**Comment by stringlytyped** [comment]
I am finally taking this PR off draft status as all major issues (test failures, typing issues, etc.) have now been resolved. Would appreciate any feedback. Since this PR is pretty substantial, I would suggest maybe starting in keylime/web/base/server.py and working outward from there

**Comment by ansasaki** [comment]
@stringlytyped Hello, I'll fill a separate PR trying to fix the certificate issues with Python 3.13 and will use the last commit you added. Please let me know if that is not ok.

**Comment by stringlytyped** [comment]
> @stringlytyped Hello, I'll fill a separate PR trying to fix the certificate issues with Python 3.13 and will use the last commit you added. Please let me know if that is not ok.

No, that's great, thanks. I realised after that the fix wasn't quite as simple as just adding the Authority Key Identifier as a number of extensions needed to be marked critical per RFC 5280, so I appreciate you sorting that

File: keylime/cert_utils.py
Comment (stringlytyped) [reply keylime/cert_utils.py:60]: Same reason as given in my [above comment](https://github.com/keylime/keylime/pull/1523#discussion_r1644570759)

File: keylime/da/record.py
Comment (stringlytyped) [reply keylime/da/record.py:49]: Yes, in fact, it is already being done on [line 473 of registrar_common.py](https://github.com/keylime/keylime/blob/bafa8801da1d1b7db0c5ce63b6bb93df557149a6/keylime/registrar_common.py#L473). When an agent is registered, there is no attestation data to record

File: keylime/cert_utils.py
Comment (ansasaki) [reply keylime/cert_utils.py:46]: OK, fair explanation. I created an issue to track this: #1559

I think the message should inform the user trying to configure Keylime which certificates are problematic so that they can decide whether to accept them or not. For the CA certificates, this would require the user to remove the malformed certificates from the trusted certificate store (and we probably should inform them about this).

For the CA certificates, the warning should appear once during initialization. For other certificates, I think it is still important to log that they are problematic. Probably the best is to allow the user to configure whether they want to ignore and accept this or not. When the user chooses to ignore this, then the messages should be suppressed.

File: keylime/da/record.py
Comment (ansasaki) [reply keylime/da/record.py:49]: Hum, OK, I thought the DA records were only for quotes received by the verifier, that is why I expected them to always have an `attestation_data`

**Comment by maugustosilva** [comment]
@mpeters @THS-on @stefanberger @ansasaki @aplanas @sergio-correia Based on discussions during the monthly video-meeting in June/24, we are ready to merge it in the next 24 hours. Please let me know if any of you see any blockers. Thanks.

**Review by maugustosilva — APPROVED** [review]

File: keylime/cert_utils.py
Comment (stringlytyped) [reply keylime/cert_utils.py:46]: @ansasaki If you could have a look at the fix I've implemented for #1559 in the latest commit, that'd be great. The code that generates the warning is on [lines 108-159 of registrar_agent.py](https://github.com/keylime/keylime/blob/7527175fdfecbe0859eb9155f41bea6c6ee7410e/keylime/models/registrar/registrar_agent.py#L108-L159). This generates a warning only when one or more certificates are added or updated in an agent record and lets the user know which certificates are at fault (in a single warning message). A new config option has been added to optionally suppress this warning or reject all problematic certs (the default is just to warn).

It looks like CodeQL is upset about these new log messages because it thinks I am logging a certificate (which is not the case). I think I'll try to just change the name of the relevant variable and see if that fixes it (let me know if you have a better idea).

I also added a template to migrate the config to include this new option. Let me know if I did this correctly or not. I ran the "convert_config" script and it didn't complain.

One last thing: Since I needed to create a new config template anyway, I used this opportunity to change the default logging format to include the request ID. When the X-Request-Id header is set in a request, its value is appended to the all log messages produced during the handling of that request. This makes it easier to see which log messages relate to which request. However, users will not notice a change until the agent is updated to set this header.

Comment (stringlytyped) [reply keylime/cert_utils.py:46]: Okay, so changing that variable name did not fix the CodeQL warning. I assume it needs to be suppressed by a maintainer.

File: keylime/models/registrar/registrar_agent.py
Comment (stringlytyped) [reply keylime/models/registrar/registrar_agent.py:152]: I tried to fix this warning by changing the variable name in [fc1217e](https://github.com/keylime/keylime/pull/1523/commits/fc1217e585322540821893a3fb7be700cadb7193) but it seems CodeQL is smarter than that (but not smart enough to realise this is a false positive). No sensitive data is being logged here: just names of the relevant certificate fields.

File: keylime/models/registrar/registrar_agent.py
Comment (stringlytyped) [reply keylime/models/registrar/registrar_agent.py:158]: Same issue as explained in my [above comment](https://github.com/keylime/keylime/pull/1523#discussion_r1655305411)

File: keylime/models/registrar/registrar_agent.py
Comment (sergio-correia) [reply keylime/models/registrar/registrar_agent.py:152]: I dismissed the CodeQL false positives.

**Comment by THS-on** [comment]
@maugustosilva I'll not be able to look fully at the changes in the next week, but @stringlytyped and I discussed the ideas and design before, so from my side there is nothing that should block merging this. As this for now only changes internals we can also easily add changes once its merged.

**Review by THS-on — APPROVED** [review]

File: keylime/models/registrar/registrar_agent.py
Comment (stringlytyped) [reply keylime/models/registrar/registrar_agent.py:152]: Thanks Sergio!

File: keylime/cert_utils.py
Comment (stringlytyped) [reply keylime/cert_utils.py]: I realised that `cert_utils.x509_der_cert` and `cert_utils.x509_pem_cert` functions are still called from other places (e.g., indirectly from the tenant) in the codebase and my fix in [commit 7527175](https://github.com/keylime/keylime/pull/1523/commits/7527175fdfecbe0859eb9155f41bea6c6ee7410e) only applies when certs are added to an agent record in the registrar. So, I've added back the warning in `cert_utils` and made copies of these functions without the warning in `models.types.Certificate`. That way, the improved warning is used when possible, and the old warning is used otherwise.

**Comment by stringlytyped** [comment]
The `basic-attestation-with-custom-certificates` test is currently failing on rawhide and CentOS Stream 10 but I don't believe this has anything to do with this PR. The OpenSSL server which is temporarily started to test revocation is producing an error that it does not recognise some CA certificate (I assume the agent's mTLS certificate?)

The error first seems to have arisen in https://github.com/keylime/keylime/pull/1557 (thanks to @kkaarreell for finding). Not sure what is causing it, but it does not seem to be a result of a recently-changed OpenSSL version

**Comment by ansasaki** [comment]
> The `basic-attestation-with-custom-certificates` test is currently failing on rawhide and CentOS Stream 10 but I don't believe this has anything to do with this PR. The OpenSSL server which is temporarily started to test revocation is producing an error that it does not recognise some CA certificate (I assume the agent's mTLS certificate?)
> 
> The error first seems to have arisen in #1563 (thanks to @kkaarreell for finding). Not sure what is causing it, but it does not seem to be a result of a recently-changed OpenSSL version

Thanks for raising this. I'll investigate what is the issue.

**Comment by stringlytyped** [comment]
> Thanks for raising this. I'll investigate what is the issue.

Excellent, thanks Anderson.

**Review by ansasaki — COMMENTED** [review]
Thanks for adding the configuration option to suppress the certificate loading warning.

About the error on the CI, from @kkaarreell findings, it is related with the `python3-requests` update. I'll continue to investigate the root cause.

File: keylime/models/registrar/registrar_agent.py
Comment (stringlytyped) [reply keylime/models/registrar/registrar_agent.py]: Ah, that's a good catch. I'll swap out that function call.

Comment (ansasaki) [reply keylime/models/registrar/registrar_agent.py]: No need to fix this right now, I'll create an issue to track this

Comment (stringlytyped) [reply keylime/models/registrar/registrar_agent.py]: It was a quick fix which I totally agree is worth the minimal effort and shouldn't break anything. So, it has been addressed in [commit a28078f](https://github.com/keylime/keylime/pull/1523/commits/a28078fb975df3b25f10b8c61d8340cea4a16a42).

**Comment by stringlytyped** [comment]
I've gone ahead and squashed the commits at Marcio's direction. If we still need to make changes before merging, I don't mind rebasing again

**Comment by maugustosilva** [comment]
Waiting for one last batch of tests, and then committing. Disclaimer, we are ignoring `CodeQL` and `fedora-rawhide` failures (i.e., these will NOT block the commits).

**Comment by ansasaki** [comment]
The test failures in Fedora Rawhide and Centos Stream 10 are caused by a regression introduced by `python-requests 2.32.3` which makes it to not load the default CA certificates anymore. This was reported upstream: https://github.com/psf/requests/issues/6730

The failures are not related with the changes in this PR.

**Comment by stringlytyped** [comment]
As mentioning commits/issues is not terribly helpful when you get the references wrong, I went ahead and updated the commit message to fix these. All tests are now passing (with the exception of the rawhide and centos failures caused by python-requests) so we should be all good to merge this PR.

Anderson: glad you managed to figure out what was causing that error. I could not make any sense of it yesterday.

**Comment by maugustosilva** [comment]
> The test failures in Fedora Rawhide and Centos Stream 10 are caused by a regression introduced by `python-requests 2.32.3` which makes it to not load the default CA certificates anymore. This was reported upstream: [psf/requests#6730](https://github.com/psf/requests/issues/6730)
> 
> The failures are not related with the changes in this PR.

Indeed. I am merging now.

**Pull request merged**
